nn/transformer.py

```diff
@@ -0,0 +1,231 @@
+"""
+Transformer Modules
+"""
+
+from __future__ import annotations
+from typing import Optional, Any, Union
+from . import relu, layer_norm, dropout, LayerRef, Module, ModuleList, Linear
+
+
+class TransformerEncoderLayer(Module):
+  """
+  Defines one layer of a standard transformer encoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, drop: float = 0.1, activation=relu,
+               layer_norm_eps: float = 1e-5, norm_first: bool = False) -> None:
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param drop: Dropout value, PyTorch name: dropout
+    :param activation: activation functional
+    :param layer_norm_eps: Epsilon value for layer normalization
+    :param norm_first: Whether to do layer norm before or afterwards
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # will change with Attention Modules
+
+    self.linear1 = Linear(dim_ff)
+    self.linear2 = Linear(dim_model)
+
+    self.activationivation = activation
+    self.norm_first = norm_first
+    self.norm_eps = layer_norm_eps
+    self.dropout = drop
+
+  def forward(self, inp: LayerRef) -> LayerRef:
+    """
+    Two possible forward variants of encoder, defined by self.norm_first
+    """
+    x = inp
+    if self.norm_first:
+      x = x + self._sa_block(layer_norm(x, epsilon=self.norm_eps))
+      x = x + self._ff_block(layer_norm(x, epsilon=self.norm_eps))
+    else:
+      x = layer_norm(x + self._sa_block(x), epsilon=self.norm_eps)
+      x = layer_norm(x + self._ff_block(x), epsilon=self.norm_eps)
+
+    return x
+
+  # self-attention block
+  def _sa_block(self, x: LayerRef) -> LayerRef:
```

Comment:
Do not describe a function by a prefixed comment. Use docstrings for that.
But anyway, if you need to describe it because the name might be ambiguous, why don't you simply extend the name such that it is clear?
```suggestion
  def _self_attention_block(self, x: LayerRef) -> LayerRef:
```

Reply:
The name + comment directly come from PyTorch, but changing the name is fine too.

Reply:
We don't need to take it over exactly char by char.